### PR TITLE
Add matcher to check whether an assertion was failed

### PR DESCRIPTION
**Pull request by abbeycode**
I added a `throwAssertion()` matcher, which returns `true` when a precondition fails or `fatalError()` is called (Issue #234). This currently only supports x86_64, and not the tvOS simulator.

Some open questions:

- [ ] ~~Fix Circle build issues~~
- [x] How do we want to credit @mattgallagher, as discussed in the issue thread?
- [x] Are we okay shipping with tvOS support based on POSIX signals? This means that tvOS schemes will have to turn off the `Debug executable` setting in their Test configuration. @mattgallagher and I discussed this in [Issue #3](https://github.com/mattgallagher/CwlPreconditionTesting/issues/3) of his repo
- [x] How do we handle non-`x86_64` test runs? I [currently have](https://github.com/Quick/Nimble/pull/248/files#diff-195a4da3a2a62ca3b233628bfb4da9a7R32) the `throwAssertion()` matcher conditionally compiling to fatally error with a clear message. Would it be better to return `false` from the matcher with a descriptive message?
- [x] Are we happy with the `throwAssertion` name? If not, what would be better?
- [x] I made CwlPreconditionTesting a submodule, and then included the files from it statically. Is this the right way to go?
- [x] Update README to document `throwAssertion()` (or whatever it ends up being called)

### CODE REVIEW

**Comment by abbeycode** [comment]
Can someone please investigate what's going on with the Circle build? It seems like it's having problems with the nested `#if` statements in ThrowAssertion.swift, but the Travis builds aren't.

**Comment by abbeycode** [comment]
After removing the tvOS checks from the code, Circle is now failing in ThrowAssertionTest.swift, which only has one `#if` statement. How is Circle configured differently than Travis?

Also, it seems that `-[PostNotificationTest testPassesWhenExpectedNotificationEventuallyIsPosted()]` fails intermittently. Is this a known issue with that test? I didn't make any changes that should have affected it.

**Comment by abbeycode** [comment]
@modocache @jeffh I'd love your feedback when you have a chance!

**Comment by jeffh** [comment]
Hey @abbeycode,

Sorry for the late response, an illness put me out for a couple weeks. I'll be busy with a conference this week. But excellent stuff in this PR!

Replies are after each question.

-------------------

> How do we want to credit @mattgallagher, as discussed in the issue thread?

Yup, add it to the README (next to the matcher) and the source files we include.

>  Are we okay shipping with tvOS support based on POSIX signals? This means that tvOS schemes will have to turn off the Debug executable setting in their Test configuration. @mattgallagher and I discussed this in Issue #3 of his repo

Does `@asm` work around this issue? I'm thinking something like:

```swift
@asm(mach_msg)
public func evil_mach_msg(msg: UnsafeMutablePointer<mach_msg_header_t>, _ option: mach_msg_option_t, _ send_size: mach_msg_size_t, _ rcv_size: mach_msg_size_t, _ rcv_name: mach_port_name_t, _ timeout: mach_msg_timeout_t, _ notify: mach_port_name_t) -> mach_msg_return_t
```

I had some limit luck using that to call fork on "unsupported" platforms before. Since we're only in tests, I think it should be ok.

> How do we handle non-x86_64 test runs? I currently have the throwAssertion() matcher conditionally compiling to fatally error with a clear message. Would it be better to return false from the matcher with a descriptive message?

I'd prefer compile errors with a clear message how to conditionally disable the expectation (I'm guessing with some swift build conditional). Otherwise, having the matcher fail with a description failure message as a fallback is acceptable. Nimble prioritizes simulator over devices in these decisions to allow CI environments to run more comprehensively.

> Are we happy with the throwAssertion name? If not, what would be better?

I don't have any other particular standout names. Sounds fine by me.

> I made CwlPreconditionTesting a submodule, and then included the files from it statically. Is this the right way to go?

Ideally some versioned dependency would be better, but CwlPreconditionTesting probably doesn't fit well with package managers. The submoduled project doesn't support Cocoapods or SPM.

While we could add the files to the podspec in the submodule, I think it might be more problematic for SPM.

SPM is strongly conventions based. I'm not sure if we could include sources that don't reside in the `Sources` directory.

This sadly make me lean towards just adding the files manually in Nimble's source, with proper attribution (README, Source Files). Please also add the SHA that the files originated from, so it'll be easier to know when we deviate from the original.

**Comment by abbeycode** [comment]
@jeffh I switched the CwlPreconditionTesting over from submodule to static inclusion, and I included instructions regarding the necessary compiler conditionals to include to avoid the `fatalError()` for non-x86_64 targets (it doesn't look like Swift currently has any notion of compiler assertions, though, so this has to remain a runtime assertion).

I didn't understand your suggestion to use `@asm(mach_msg)` for tvOS compatibility. How is that supposed to work?

Also, I'm bewildered by the continuing Circle and Travis build failures. Could you please take a look?

**Comment by abbeycode** [comment]
@jeffh Please let me know how to fix the build failures when you have a chance.

**Comment by jeffh** [comment]
I think the issue is that #endif needs a newline or the compiler might not
like having a build conditional over the entire file,, but I'm not entirely
sure.

**Comment by jeffh** [comment]
Also, for the `@asm`. You can use `@asm` like a forward declaration:

```swift
@asm(mach_msg)
public func evil_mach_msg(msg: UnsafeMutablePointer<mach_msg_header_t>, _ option: mach_msg_option_t, _ send_size: mach_msg_size_t, _ rcv_size: mach_msg_size_t, _ rcv_name: mach_port_name_t, _ timeout: mach_msg_timeout_t, _ notify: mach_port_name_t) -> mach_msg_return_t
```

Now you can use `evil_mach_msg` instead, which the compiler won't complain about.

```swift
// code that needs to use `mach_msg`
evil_mach_msg(...)
```

That is predicated on the fact that the asm code can still be linkable. Sorry, I should have elaborated that more.

**Comment by abbeycode** [comment]
@jeffh Thanks for the explanation; I now understand what the `@asm` attribute is, and how to use it. I'm not sure I follow how it helps with the question I asked, though:

> Are we okay shipping with tvOS support based on POSIX signals? This means that tvOS schemes will have to turn off the Debug executable setting in their Test configuration. @mattgallagher and I discussed this in Issue #3 of his repo

Would the approach you described allow tvOS to use the primary mach-based approach, instead of relying on the POSIX approach?

**Comment by jeffh** [comment]
Yeah, it theoretically should work on this, because it bypasses the
compiler's unavailable attribute.

**Comment by abbeycode** [comment]
@jeffh I'm trying out the `@asm` attribute, but it's not working (in Xcode 7.2 or 7.3b5). Both are giving me this error:

> Unknown attribute 'asm'

Is there any import or compiler flag I'm missing or something? Also, looking at the `mach_msg` definition in message.h, I see a `__TVOS_PROHIBITED` annotation. Even if we get `@asm` to compile, won't that be a problem? Or are you saying that it prevents compilation, but the function will still be there are runtime?

**Comment by morganchen12** [comment]
Try using `@asmname("mach_msg")`.

By using `@asmname`, we're basically betting on the function being available at runtime even though the `__TVOS_PROHIBITED` annotation says we're not allowed to use it.

**Comment by abbeycode** [comment]
@morganchen12 I get the same thing (in Xcode 7.3b5):

> Unknown attribute 'asmname'

**Comment by briancroom** [comment]
@abbeycode @morganchen12 I imagine this is the cause of the error: https://github.com/apple/swift/commit/fbd2e4d872d1aa57bfba2ab1f4d280bb1e90cbb8

I wonder if it would be possible instead to use a C source file to re-export the function under a different name? I'm certain exactly what that would need to look like though.

**Comment by abbeycode** [comment]
@briancroom @morganchen12 @jeffh Some questions, in light of the discussion above:

1. Are we sure, before taking the time to try getting that to work, that `mach_msg` is definitely available in tvOS's runtime? It would suck to make it callable, then find out that it's disallowed _because_ it's not there
2. Is hacking around compile-time limitations the best approach to solving this? I have it working now using POSIX signals instead of mach messages, and it works. It just has the slight drawback that for tvOS targets using the new call, they would need to turn off a single scheme option (Debug executable)
3. If we go with the POSIX approach, though, I would need to figure out how to get the SPM packaging to work, since each approach uses an alternate implementation of `BadInstructionException`. In the Xcode project, there's no problem, as `CwlCatchBadInstruction.(h/m/swift)` belong only to the OS X and iOS targets and `CwlCatchBadInstructionPOSIX.swift` belongs to only tvOS, but I don't see a way to exclude certain source files from SPM
4. Given 2 and 3 above, are we better off perhaps leaving out tvOS support altogether for this feature, creating a separate issue to add it later?

**Comment by morganchen12** [comment]
1. Yes, although even if it is there in the runtime there may be a good reason (even in unit tests on simulator) why it's labeled as prohibited.
2. Probably not! Giving up the ability to debug a unit test run with `lldb` is pretty limiting, though. Signal handlers are also very hard to work with.

I'm not sure about 3 or 4.

**Comment by abbeycode** [comment]
Alright, I removed tvOS and Swift Package Manager support, and got the build (mostly) fixed. The single build error that's showing up on Circle now is one that comes up intermittently in Xcode as well, and can always be fixed by a clean/rebuild.

**Comment by briancroom** [comment]
This is a really interesting error happening on CircleCI, @abbeycode. Do you have any theories about what is triggering it? It seems like it could be a race condition between the production of the `Nimble-Swift.h` header, and the compilation of the `.m` file that imports it. Unfortunately the CircleCI build failed a second time, even after I kicked it off again.

I would be hesitant to merge this if it would make our CI builds produce false negatives on a regular basis. I wonder reliability could be improved by manually adding the necessary declarations into `CwlCatchBadInstruction.m` instead of importing the generated header?

**Comment by abbeycode** [comment]
@briancroom I was thinking along the same lines as you are, and agree it's a little risky introducing that kind of instability. It looks like it may have cleared up on Circle, though. There was a successful build before I pushed (not sure how that one was triggered), and the latest build succeeded. but now the iOS build on Travis has failed. Argh.

Regarding the workaround you proposed, I was trying to leave the CwlPreconditionTesting sources unmodified, which is the only reason I'm hesitant. It would make it easier and less error prone in the future to just dump the latest files wholesale, instead of doing a diff to make sure we keep any modifications.

**Comment by jeffh** [comment]
TL;DR: I think it's ok to **not support tvOS** for now.

I have an unsupported assumption that `mach_msg` wouldn't be broken in tvOS since mach messaging is pretty fundamental to the Darwin kernel. Many Apple Frameworks that use IPC/XPC will implicitly rely on this function to work properly. I'm not saying it wouldn't break, but I'm just hedging bets ;)

For **3**, SPM supposedly has an [exclude parameter](http://kostiakoval.github.io/posts/exclude-folders---new-feature-in-swift-package-manager/) to exclude folders.

For **4**, let's table tvOS support for this feature. I'm perfectly fine with that.

---

As for the build exception, I'm seeing this locally on my machine. I'm guessing Swift is name-mangling `BadInstructionException` when exposing it to Objective-C, it's better to be explicit about it's name:

```swift
@objc(BadInstructionException)
public class BadInstructionException: NSException {
...
}
```

That seems to resolve the build error on my machine.

**Comment by jeffh** [comment]
Ignore my previous comment about the `@objc`, it doesn't seem to work after removing derived data. I'll play around with it a bit more.

**Comment by abbeycode** [comment]
Ok, I also created [an issue](https://github.com/mattgallagher/CwlPreconditionTesting/issues/5) on the CwlPreconditionTesting project, in case @mattgallagher has any ideas.

**Comment by jeffh** [comment]
Well it gets more interesting, but it's basically seems like a non-deterministic generation of `Nimble-Swift.h` based on architecture.

So if we modify `CwlBadInstructionException.swift` to add an else-generation:

```swift
#if arch(x86_64)
// same code as before
#else
    @objc public class BAAAAADCompiler: NSException { }
#endif
```

If you look at different runs of generating `Nimble-Swift.h` interface header, sometimes you'll see it follow the x86_64 build conditional and sometimes not:

```swift
// If the compiler decided to generate as non-x86-64
SWIFT_CLASS("_TtC6Nimble15BAAAAADCompiler")
@interface BAAAAADCompiler : NSException
- (nonnull instancetype)initWithName:(NSString * _Nonnull)aName reason:(NSString * _Nullable)aReason userInfo:(NSDictionary * _Nullable)aUserInfo OBJC_DESIGNATED_INITIALIZER;
- (nullable instancetype)initWithCoder:(NSCoder * _Nonnull)aDecoder OBJC_DESIGNATED_INITIALIZER;
@end
```

Versus sometimes the compiler will produce:

```swift
// If the compiler decided to generate as x86-64
SWIFT_CLASS("_TtC6Nimble23BadInstructionException")
@interface BadInstructionException : NSException
- (nullable instancetype)initWithCoder:(NSCoder * _Nonnull)aDecoder OBJC_DESIGNATED_INITIALIZER;
+ (kern_return_t)catch_mach_exception_raise_state:(mach_port_t)exception_port exception:(exception_type_t)exception code:(mach_exception_data_type_t const * _Null_unspecified)code codeCnt:(mach_msg_type_number_t)codeCnt flavor:(int32_t * _Null_unspecified)flavor old_state:(natural_t const * _Null_unspecified)old_state old_stateCnt:(mach_msg_type_number_t)old_stateCnt new_state:(thread_state_t _Null_unspecified)new_state new_stateCnt:(mach_msg_type_number_t * _Null_unspecified)new_stateCnt;
@end
```

As a work-around, it might be better to have a build conditional that produces a runtime error / exception instead of conditionally generating this class.

**Comment by abbeycode** [comment]
Should we file a Swift bug?

**Comment by jeffh** [comment]
Yeah! :+1: In theory, this could just require one file to repro. I happened to move on to other things ATM, but feel free to file one

**Comment by abbeycode** [comment]
@jeffh Thinking about it more, is it because it's building for `arm` and `x86` at the same time, and the `arm` build is failing? It may not be a bug in Swift at all, then, and the right answer is to raise a runtime error, and the error would be what you suggested earlier: make sure it builds for all architectures, but have it fail at runtime for non-x86.

**Comment by benjaminsnorris** [comment]
I am really excited for the possibilities this would open up. It feels frustrating right now to not be able to test the edge cases of the app to define the cases where it should fail. Hopefully this can be merged soon!
